### PR TITLE
add rule no-empty-array

### LINKED ISSUE #199: Rule S4030: Empty collections should not be accessed or iterated
Implements [RSPEC-4158](https://jira.sonarsource.com/browse/RSPEC-4158)

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
Fixes #199

### CODE REVIEW

File: sonarts-core/docs/rules/no-empty-array.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+## Noncompliant Code Example
+
+```typescript
+let strings = [];
+
+if (strings.includes("foo")) {}  // Noncompliant
+
+for (String str of strings) {}  // Noncompliant
+
+strings.forEach() // Noncompliant
```
Comment (vilchik-elena) [inline sonarts-core/docs/rules/no-empty-array.md]: let's add some argument

File: sonarts-core/src/rules/noEmptyArrayRule.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+  public static MESSAGE = "Remove this call; the collection can only be empty here.";
+
+  public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): tslint.RuleFailure[] {
+    const visitor = new Visitor(this.getOptions().ruleName, program);
+    const symbols = SymbolTableBuilder.build(sourceFile, program);
+    const collectionSymbols = getCollectionSymbols(symbols, program);
+    collectionSymbols
+      // filter out array declared as fields
+      .filter(symbolAndDeclaration => !ts.isPropertyDeclaration(symbolAndDeclaration.declaration.parent))
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noEmptyArrayRule.ts]: remove usage of `ts.is...`

File: sonarts-core/src/rules/noEmptyArrayRule.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+      // keep only symbols initialized to empty array literal or not initialized at all
+      .filter(symbolAndDeclaration => {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noEmptyArrayRule.ts]: I would extract this filter logic

File: sonarts-core/src/rules/noEmptyArrayRule.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+          }),
+      );
+    return visitor.getIssues();
+  }
+
+  private static isConcatCall(node: ts.Node) {
+    return isPropertyAccessExpression(node.parent) && node.parent.name.text === "concat";
+  }
+
+  private static isNewEmptyCollectionCreation(node: ts.Node): boolean {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noEmptyArrayRule.ts]: -> `isEmptyCollection`

File: sonarts-core/src/symbols/collectionUtils.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  isElementWrite,
+  isVariableWrite,
+  isWritingMethodCall,
+];
+
+/**
+ * Returns an array of pair symbol-declaration storing collections.
+ * Parameters, exported/imported and type related symbols are filtered out.
+ */
+export function getCollectionSymbols(symbols: SymbolTable, program: ts.Program) {
```
Comment (vilchik-elena) [inline sonarts-core/src/symbols/collectionUtils.ts]: all logic here is so specific (so many things are filtered out). Can we move this class to `rules` dir?

File: sonarts-core/tests/rules/noEmptyArrayRule/noEmptyArrayRule.lint.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+var arrayWrittenInsideArrow2 = [];
+func(n => arrayWrittenInsideArrow2 = otherArray);
+arrayWrittenInsideArrow2[1]; // OK
+
+
+// Interface Declaration
+
+interface Array<T> {
+  equals(array: Array<T>): boolean // OK
```
Comment (vilchik-elena) [inline sonarts-core/tests/rules/noEmptyArrayRule/noEmptyArrayRule.lint.ts]: what do we test here?

File: sonarts-core/tests/rules/noEmptyArrayRule/noEmptyArrayRule.lint.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+export type MyArray = T[];
+
+export namespace MyArray {
+  export function is(value: any): value is MyArray {
+    return value && Types.isArray(value);
+  }
+}
+
+function foo() {
```
Comment (vilchik-elena) [inline sonarts-core/tests/rules/noEmptyArrayRule/noEmptyArrayRule.lint.ts]: name function meaningfully

File: sonarts-core/tests/rules/noEmptyArrayRule/noEmptyArrayRule.lint.ts (hunk 6 above)
Comment (andrea-guarino-sonarsource) [reply sonarts-core/tests/rules/noEmptyArrayRule/noEmptyArrayRule.lint.ts]: This is to test that we're filtering out symbols (`Array in this case`) whose declaration is an InterfaceDeclaration. I will better document this.

File: sonarts-core/src/rules/noEmptyArrayRule.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    description: "Empty collections should not be accessed or iterated",
+    rationale: tslint.Utils.dedent``,
+    optionsDescription: "",
+    options: null,
+    rspecKey: "RSPEC-4158",
+    type: "functionality",
+    typescriptOnly: false,
+  };
+
+  public static MESSAGE = "Remove this call; the collection can only be empty here.";
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noEmptyArrayRule.ts]: Review this usage of '...' as it can only be empty here

**Pull request merged**